### ISSUE TITLE
PPX not found when building in an alpine docker container

### ISSUE THREAD

**Issue by johnhaley81** [baransu/graphql_ppx_re#44]
When using this lib in a docker container built from alpine you'll get errors saying things like

`Command line: /app/node_modules/@baransu/graphql_ppx_re/ppx '/tmp/camlppx375570' '/tmp/camlppx57df4b'`

and

```
FAILED: packages/worker/src/Foo.reast
/app/node_modules/bs-platform/lib/bsc.exe  -w -30-40+6+7+27+32..39+44+45+101+A-9-40-42-48-102 -warn-error +A-3-44 -color always -bs-jsx 3 -ppx /app/node_modules/@baransu/graphql_ppx_re/ppx  -o packages/worker/src/Foo.reast -bs-syntax-only -bs-binary-ast /app/packages/worker/src/Foo.re
sh: /app/node_modules/@baransu/graphql_ppx_re/ppx: not found
```

Switching to a container that's not built on Alpine fixes the issue.

**Comment by baransu** [comment]
Does the post install script copied ppx binary in place? Is the file available and in correct place?

**Comment by johnhaley81** [comment]
@baransu it is and even when I go into the docker container and run it manually I get that same error.

**Comment by maxkorp** [comment]
So the outputted `ppx` file is there, but not executable (it does have executable permissions) because of the lack of glibc. 

We were able to fix this for our project for now, by installing glibc into our alpine image by adding the following to our dockerfile (I still have to check and see if there are newer glibc versions to use instead). 
```
RUN wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
RUN wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-2.28-r0.apk
RUN apk add --no-cache glibc-2.28-r0.apk
```

The reason it worked with the original mhallin ppx was because that was built in alpine already using opam (It didn't have any dynamic deps specific to alpine either so it worked in "vanilla" linux as well). When building via esy (at least esy installed via npm), glibc is required at build time at least. I did not check if it was required at run time when built that way, or if a version of esy built in alpine without glibc (if such a thing is reasonably feasible) would work outside of alpine. 

The following is only really relevant if building it with glibc in docker works both in "vanilla" linux and in alpine-without-glibc, but I'm including it in case it is, because then the build/deploy pipelines can be updated to provide a linux version that works everywhere without the additional step of installing glibc (which adds a noticeable % increase to the docker image size). 

The esy devs seem to recommend using esy-docker (https://github.com/andreypopp/esy-docker) as a baseline for building things via esy in docker. Note that the project is behind a bit (e.g. node 10 and esy 0.4.3), and is also somewhat complicated (it uses some makefile and config to add several discrete dockerfiles that expect a certain project structure and more. I took the output from there and tweaked it to use more current versions of (almost) everything, and added a script to go with it to output alpine compatible builds (that again may or may not require glibc at which point there is no benefit if they do)  and output them locally

Dockerfile
```
# Inspired by https://github.com/andreypopp/esy-docker

FROM node:12.13-alpine as build

ENV TERM=dumb LD_LIBRARY_PATH=/usr/local/lib:/usr/lib:/lib

RUN mkdir /esy
WORKDIR /esy

ENV NPM_CONFIG_PREFIX=/esy
RUN npm install -g --unsafe-perm esy@latest

# now that we have esy installed we need a proper runtime

FROM alpine:3.8 as esy

ENV TERM=dumb LD_LIBRARY_PATH=/usr/local/lib:/usr/lib:/lib

WORKDIR /

COPY --from=build /esy /esy

RUN apk add --no-cache ca-certificates wget bash curl perl-utils git patch gcc g++ musl-dev make m4

RUN wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
RUN wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-2.28-r0.apk
RUN apk add --no-cache glibc-2.28-r0.apk

ENV PATH=/esy/bin:$PATH

# gql stuff

RUN git clone https://github.com/baransu/graphql_ppx_re /app

WORKDIR /app
RUN mkdir -p /builds

RUN esy install
RUN esy build
```

build.sh
```
mkdir -p output
docker build -t gql-ppx-alpine-builder .
docker run --mount type=bind,src=$(pwd)/output,dst=/builds gql-ppx-alpine-builder cp _build/default/src/bucklescript_bin/bin.exe /builds/bin.exe
```

**Comment by maxkorp** [comment]
Well poop, Building it that way does _not_ work in alpine without glibc, so that is precisely 0% helpful :/

**Comment by maxkorp** [comment]
Ok well I get some other oddball error using it via that process, probably because I'm only copying in that binary and not anything else 
`Fatal error: exception Failure("Ast_mapper: OCaml version mismatch or malformed input")`

**Comment by baransu** [comment]
@johnhaley81 @maxkorp I just released v0.4.9. It’s build using alpine image and should work on alpine and other common distributions. I tested it using `ldd`. Please let me know if it resolves your problem 🙂

**Comment by jfrolich** [comment]
Awesome work! Just curious. Must we always ship compiled PPXs for each architecture? Seems like quite a hassle to maintain for each PPX. Is there no better way like shipping bytecode (or compiling as a postinstall step or so).

**Comment by baransu** [comment]
I'm not sure about bytecode runtime, but with post-install compile you would need while `esy` env with `ocaml` and whatnot. A few minutes and few GB required to test ReasonML with GraphQL. 

We probably need only 3 binaries, each one for one major OS (generic Linux - solved by compiling on alpine with static linking, macOS and Windows).

**Comment by baransu** [comment]
Closing the issue as it's resolved since version `0.4.9`. Please create a new issue ore reopen if it's still a problem.

**Comment by maxkorp** [comment]
Just stopping in to confirm that this is in fact working for me locally. Thanks!
